conda_declarative/edit.py

```diff
@@ -76,20 +80,86 @@ class SortOrder(Enum):
     DESC = "descending"
 
 
+class SpinnerWidget(Static):
+    """Textual widget which displays a spinner."""
+
+    DEFAULT_CLASSES = "hidden"
+    DEFAULT_CSS = """
+    SpinnerWidget {
+        visibility: visible;
+    }
+    SpinnerWidget.hidden {
+        visibility: hidden;
+    }
+    """
+
+    def __init__(self):
+        super().__init__("")
+        self._spinner = Spinner("dots")
+
+    def on_mount(self) -> None:
+        """Set the update interval of the spinner."""
+        self.update_render = self.set_interval(1 / 60, self.update_spinner)
+
+    def update_spinner(self) -> None:
+        """Update the spinner status."""
+        self.update(self._spinner)
+
+    def set_text(self, text: str) -> None:
+        """Set the text next to the spinner.
+
+        Parameters
+        ----------
+        text : str
+            Text to display
+        """
+        self._spinner.update(text=text)
+
+    def hide(self):
+        """Hide the spinner."""
+        self.add_class("hidden")
+
+    def show(self):
+        """Show the spinner."""
+        self.remove_class("hidden")
+
+
+class HidableProgressBar(ProgressBar):
+    """A progress bar that can be easily hidden."""
+
+    DEFAULT_CLASSES = "hidden"
+
+    def hide(self):
+        """Hide the progress bar."""
+        self.add_class("hidden")
+
+    def show(self):
+        """Show the progress bar."""
+        self.remove_class("hidden")
+
+
 class EditApp(App):
     """Main application which runs upon `conda edit`."""
 
     BINDINGS = [
         Binding("ctrl+q", "quit", "Quit", priority=True),
         Binding("ctrl+s", "save", "Save", priority=True),
+        Binding("ctrl+e", "sync", "Sync", priority=True),
     ]
 
     DEFAULT_CSS = """
+    HidableProgressBar {
+        visibility: visible;
+        border: solid red;
+    }
+    HidableProgressBar.hidden {
+        visibility: hidden;
+    }
     #output {
         height: 1fr;
     }
     #progress-area {
-        height: 1;
+        height: 2;
```

Comment:
I think this is a little bit small. I can't really see any details in this space
![Screenshot from 2025-07-09 11-22-35](https://github.com/user-attachments/assets/1a861a2e-d7fc-4f5c-a34a-094a91d9f322)

Reply:
Hmm, this is pretty weird - is the progress bar only partially visible on your terminal? On mine it takes up a full row, so is visible just above the bottom Footer element.

Reply:
Ya, the spineer is only partially visibile in my terminal. That's in the built in vscode terminal. I also tried in [terminator](https://gnome-terminator.org/).

Reply:
Whoo this one took me a while to understand, but it should be sorted out now. In short:

Conda reports (download) progress using the CondaReporterBackend plugin hooks. Importantly, it can spawn multiple progress bars to be shown simultaneously, which each have to be handled on the TUI. I ended up reusing the apprach taken by `textual` to implement the toast notification system to create a popup which displays these progress bars as needed when the user decides to apply the changes.

Reply:
woah, testing this out, the new ui is rad :surfer: Nice one!